### ISSUE TITLE
Installing `rJava` on arm64 macOS 

### ISSUE THREAD

**Issue by pat-s** [s-u/rJava#287]
I have installed the CRAN arm64 R binary.

I am trying to install `rJava` but fail to do so both with the binary and installing from source.

I have `openjdk@11` installed via homebrew and ran `R CMD javareconf`.

My `.R/Makevars` looks as follows

```
LDFLAGS  +=-L/opt/homebrew/opt/jpeg/lib -L/opt/homebrew/opt/libpng/lib  -L/opt/homebrew/opt/pcre2/lib
CFLAGS   = -I/opt/homebrew/include
CPPFLAGS = -I/opt/homebrew/include
CXXFLAGS = -I/opt/homebrew/include
```

Any help is appreciated.

## Source

```r
ccache clang -o libjri.jnilib Rengine.o jri.o Rcallbacks.o Rinit.o globals.o rjava.o  -dynamiclib -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/opt/homebrew/Cellar/openjdk@11/11.0.12/libexec/openjdk.jdk/Contents/Home/lib/server -ljvm -L/opt/R/arm64/lib -F/Library/Frameworks/R.framework/.. -framework R -L/opt/R/arm64/lib -lpcre2-8 -llzma -lbz2 -lz -licucore -ldl -lm -liconv
ld: library not found for -lpcre2-8
```

I've tried various attempts to tell R where `pcre2` can be found but without success. Also it `pcre2` is installed with v10 whereas it seems `rjava` looks for v8 specifically?

## Binary

```r
library(rJava)
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.

Error: package or namespace load failed for ‘rJava’:
 .onLoad failed in loadNamespace() for 'rJava', details:
  call: fun(libname, pkgname)
  error: JVM could not be found
In addition: Warning messages:
1: In system("/usr/libexec/java_home", intern = TRUE) :
  running command '/usr/libexec/java_home' had status 1
2: In fun(libname, pkgname) :
  Cannot find JVM library 'NA/lib/server/libjvm.dylib'
Install Java and/or check JAVA_HOME (if in doubt, do NOT set it, it will be detected)
```

Here I am not sure why JVM cannot be detected as its installed. I assume it might look for the x86 binary? If so, support for the arm64 binary would be great.

**Comment by pat-s** [comment]
```sh
brew install openjdk@11
sudo ln -sfn /opt/homebrew/opt/openjdk@11/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk-11.jdk
sudo R CMD javareconf

R -q -e "install.package('rJava')"
```

**Comment by pat-s** [comment]
Hitting the error again, this time on a **fresh install** of R 4.1.2 (CRAN).

Ran `sudo R CMD javareconf` and `rJava` is still not being able to find `-lpcre2-8` although it see it being present in `/opt/R/arm64/lib` and `/opt/R/arm64/include`.

```
clang -arch arm64 -o libjri.jnilib Rengine.o jri.o Rcallbacks.o Rinit.o globals.o rjava.o  -dynamiclib -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/opt/homebrew/Cellar/openjdk@11/11.0.12/libexec/openjdk.jdk/Contents/Home/lib/server -ljvm -L/opt/R/arm64/lib -F/Library/Frameworks/R.framework/.. -framework R -L/opt/R/arm64/lib -lpcre2-8 -llzma -lbz2 -lz -licucore -ldl -lm -liconv
ld: library not found for -lpcre2-8
ld: library not found for -lpcre2-8
```

I've been playing around with `LDFLAGS` and similar but they do not seem to honored.

Help is appreciated.

**Comment by s-u** [comment]
@pat-s those flags have nothing to do with rJava, they come from the R installation as you can check with

```
$ R CMD config --ldflags
 -L/opt/R/arm64/lib -F/Library/Frameworks/R.framework/.. -framework R -L/opt/R/arm64/lib -lpcre2-8 -llzma -lbz2 -lz -licucore -ldl -lm -liconv
```

If you want to embed R then you will need `pcre2` and `xz` libraries installed. You can download them from https://mac.r-project.org/libs-arm64/ or you can use [recipes](https://github.com/R-macos/recipes) to build them.

If you don't need to embed R then use `--disable-jri` in which case you only need Java.

That said, the recommended way is to install rJava from CRAN since that saves you the hassle.

**Comment by pat-s** [comment]
> That said, the recommended way is to install rJava from CRAN since that saves you the hassle.

I am doing so - my goal is to install from source (for various reasons). I am aware that I could use the binary (which works fine) but this won't work on R-devel.

I see `prce` being present in `/opt/R/arm64/lib` and `/opt/R/arm64/include` after a R CRAN installation (just fyi I am installing with `brew install --cask r` not `brew install r`).
I just found that while `pcre` is present, `pcre2` is missing.
Downloading [pcre2 from libs-arm64](https://mac.r-project.org/libs-arm64/pcre2-10.34-darwin.20-arm64.tar.gz) and moving it into `/opt/R/arm64/lib` solves it and source installation succeeds.

I am wondering: What is the reason to not include `pcre2` by default within the CRAN installation? This would prevent such issues for users.
I am seeing several header files being shipping with the installation in `/opt/R/arm64/lib` and their size is very small.
Even as a dev (and being aware of `libs-arm64`) this solution path was not clear to me.

Also I think it would be great if during the `rJava` installation `LDFLAGS` could be picked up so that one could also make use of homebrew's `pcre2` installation. Maybe this could even be a default search path?

So in summary, my wish would be for a more trouble-free/smart installation of `rJava`, both from source with R-release and R-devel.

Thanks for your work and help.

**Comment by s-u** [comment]
The CRAN R build is static, so users never need those libraries as they are part of the resulting binary, so there is no point in shipping them. As long as you stick to CRAN binaries everything "just works". I can't speak for what Homebrew is doing - according to your report it doesn't declare correct dependencies if I read it correctly.

rJava doesn't care - it only uses what R tells it, rJava doesn't use any of those libraries, it's simply asks R how to link to `libR` for embedding. In the static case R the linking has to include all libraries since there is no dynamic loader to resolve dependencies _for R_. If you think the flags are wrong, fix them in R - they must match how R was built, you can't just use other libraries.

Again, if you just want the package,  use `--disable-jri` since all of the above only affects JRI, i.e., if you want to create a Java program that embeds R.
